### PR TITLE
[FIX] fields: prevent writing float fields if not changed

### PR DESCRIPTION
**Pull request by poma-odoo**
[FIX] fields: prevent writing float fields if not changed

 In some rare cases, this line causes float fields to write even when the value has not been changed because of precision. This causes bugs like #91626
Here is an example, when record value is 0.12:

origin[name]=0.12 and cache is calculated here:

```python
class Float(Field):
    #...
    def convert_to_cache(self, value, record, validate=True):
        # apply rounding here, otherwise value in cache may be wrong!
        value = float(value or 0.0)
        if not validate:
            return value
        digits = self.get_digits(record.env)
        return float_round(value, precision_digits=digits[1]) if digits else value
```
the value in cache is:
```
>>>float_round(0.12, precision_digits=4)
0.12000000000000001
```
then it is detected as a changed field because `0.12 != 0.12000000000000001`

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by poma-odoo** [comment]
Here is another fix:
```python
class Float(Field):
    def convert_to_column(self, value, record, values=None, validate=True):
        result = float(value or 0.0)
        digits = self.get_digits(record.env)
        if digits:
            precision, scale = digits
            result = float_repr(float_round(result, precision_digits=scale), precision_digits=scale)
        return result

    def convert_to_cache(self, value, record, validate=True):
        # apply rounding here, otherwise value in cache may be wrong!
        value = float(value or 0.0)
        if not validate:
            return value
        # HERE
        # - digits = self.get_digits(record.env)
        # - return float_round(value, precision_digits=digits[1]) if digits else value
        return float(self.convert_to_column(value, record, validate))
```

**Comment by kbr-odoo** [comment]
@poma-odoo Hello, Pouya!
Do you believe we'll be able to progress with this fix soo-ish?
Thank you!

File: odoo/fields.py [hunk 1]
```diff
@@ -2844,8 +2844,16 @@ def convert_to_write(self, value, record):
...
+            def compare_with_cache(record, origin, name):
+                if record._fields[name].type == 'float':
+                    return float_compare(record[name], origin[name],
+                                             precision_digits=record._fields[name].get_digits(record.env)[1])
+                elif record._fields[name].type == 'monetary':
+                    return record[record._fields[name].currency_field].compare_amounts(record[name], origin[name])
+                elif isinstance(record[name], BaseModel):
+                    return record[name]._origin != origin[name]
+                else:
+                    return record[name] != origin[name]
```
[the remark this answers is not in the document]
Comment (VincentSchippefilt) [reply odoo/fields.py:2856]: As a matter of fact, this method should be defined in each field type:
- float and monetary for their special cases
- relational fields for the `_origin` case
- field for the fallback case

**Review by VincentSchippefilt — CHANGES_REQUESTED** [review]
This change requires some tests showing the failing cases of the previous implementation.

**Pull request closed without merging**